rfcs/text/0428-extending-and-duplicating-types.md

```diff
@@ -0,0 +1,970 @@
+  "type": "object",
+  "title": "Country",
+  "properties": {
+    "https://blockprotocol.org/@alice/property-type/name": {
+      "$ref": "https://blockprotocol.org/@alice/property-type/name/v/1"
+    }
+  },
+  "required": ["https://blockprotocol.org/@alice/property-type/name"]
+}
+```
+
+Entity type `Region` with `Country` as a supertype
+
+```json
+{
+  "$schema": "https://blockprotocol.org/types/modules/graph/0.3/schema/entity-type",
+  "kind": "entityType",
+  "$id": "https://blockprotocol.org/@alice/entity-type/region/v/1",
+  "type": "object",
+  "title": "Region",
+  "allOf": [
+    { "$ref": "https://blockprotocol.org/@alice/entity-type/country/v/1" }
+  ],
+  "properties": {
+    "https://blockprotocol.org/@alice/property-type/blurb": {
+      "$ref": "https://blockprotocol.org/@alice/property-type/blurb/v/1"
+    }
+  },
+  "required": ["https://blockprotocol.org/@alice/property-type/blurb"]
+}
+```
+
+The second, cyclic version of the `Country` entity type
+
+```json
+{
+  "$schema": "https://blockprotocol.org/types/modules/graph/0.3/schema/entity-type",
+  "kind": "entityType",
+  // Because of versioning, we cannot change this version to /v/1 and create a "proper" cycle.
+  "$id": "https://blockprotocol.org/@alice/entity-type/country/v/2",
+  "type": "object",
+  "title": "Country",
+  "allOf": [
+    { "$ref": "https://blockprotocol.org/@alice/entity-type/region/v/1" }
+  ],
+  "properties": {
+    "https://blockprotocol.org/@alice/property-type/name": {
+      "$ref": "https://blockprotocol.org/@alice/property-type/name/v/1"
+    },
+    "https://blockprotocol.org/@alice/property-type/location": {
+      "$ref": "https://blockprotocol.org/@alice/property-type/location/v/1"
+    }
+  },
+  "required": ["https://blockprotocol.org/@alice/property-type/name"]
+}
+```
+
+This sort of type hierarchy should _not_ be encouraged within the type extension system, as the circular dependencies make types difficult to reason about. While the type hierarchy might be completely valid (as it would be in this case), we should safeguard users from making redundant type structures that look like the above.
+
+In this specific contrived example, creating a new entity type based on `Region` instead of a new version of `Country` might even encode semantic meaning better than re-defining `Country`.
+
+## Multiple supertypes - checking compatibility
+
+As described in the [Guide-level explanation](#multiple-supertypes), when extending multiple entity types, they must be able to coexist in a "compatible manner", which means that the entity types' `properties` and `links` comply with the following:
+
+- For each property (base URL on the top level of `properties`) that exists in multiple entity types:
+
+  - all entity types refer to the same versioned URL of the property type **or** compatible versions of the property type (through the [versioning RFC](./0408-versioning-types.md#determining-type-compatibility) definition of 'compatible')
```

Comment:
This is perhaps the biggest open question to me now. I wonder if structural compatibility is good enough, maybe it is.

I have concerns, especially in the current system where we don't have fine-grained constraints on values, that the semantic meaning of property types is important. What if I had a "Column ID" field in one of my entity types that I use in my API, and it's simple a `Text` value, with a `description` that says "must be the ID of a specific account". And then what happens if a new version of "Column ID" is published which modifies the description to be "must be the ID of a specific account within Europe".

This is perhaps one of the compromises we have to be happy with, because not everything will be able to be rigorously encoded, and there needs to be some leeway whereby things can be updated with ease. Relying on everything having the same version of a type sounds like a maintenance nightmare if you don't manage all the types, and you might be waiting on some upstream creator to update their type to have the new version of `Name` which fixed a typo in the description.. which would just encourage forking.

Reply:
The requirements about compatibility have been completely removed in the new version of the doc. The relevant section is in the rationale and alternatives, and describes the difficulty of enforcing satisfiable schemas.